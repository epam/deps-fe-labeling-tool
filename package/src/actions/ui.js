import { createAction } from 'redux-actions'

const FEATURE_NAME = 'UI'

const setMarkupObjectsFilter = createAction(
  `${FEATURE_NAME}/SET_MARKUP_OBJECTS_FILTER`
)

const setAssignToFieldsFilter = createAction(
  `${FEATURE_NAME}/SET_ASSIGN_TO_FIELDS_FILTER`
)

const setExpandedListKeys = createAction(`${FEATURE_NAME}/SET_KEYS`)

const toggleExpandedListKey = createAction(`${FEATURE_NAME}/TOGGLE_KEY`)
const setActiveSidebar = createAction(`${FEATURE_NAME}/SET_ACTIVE_SIDEBAR`)

const addTemporaryFieldIndex = createAction(`${FEATURE_NAME}/ADD_TEMPORARY_FIELD`)
const deleteTemporaryFieldIndex = createAction(`${FEATURE_NAME}/DELETE_TEMPORARY_FIELD_INDEX`)

const resetDefault = createAction(`${FEATURE_NAME}/RESET_DEFAULT`)

export {
  setMarkupObjectsFilter,
  setActiveSidebar,
  setExpandedListKeys,
  toggleExpandedListKey,
  addTemporaryFieldIndex,
  setAssignToFieldsFilter,
  deleteTemporaryFieldIndex,
  resetDefault
}
