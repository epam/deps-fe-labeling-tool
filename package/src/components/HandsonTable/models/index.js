
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
