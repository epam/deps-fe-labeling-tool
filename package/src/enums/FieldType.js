const FieldType = {
  CHECKMARK: 'checkmark',
  ENUM: 'enum',
  LIST: 'list',
  PAIR: 'pair',
  STRING: 'string',
  TABLE: 'table',
  DATE: 'date'
}

const RESOURCE_FIELD_TYPE = {
  [FieldType.STRING]: 'String',
  [FieldType.ENUM]: 'Enumeration',
  [FieldType.CHECKMARK]: 'Checkbox',
  [FieldType.LIST]: 'List',
  [FieldType.PAIR]: 'Key Value Pair',
  [FieldType.TABLE]: 'Table',
  [FieldType.DATE]: 'Date'
}

export {
  FieldType,
  RESOURCE_FIELD_TYPE
}
