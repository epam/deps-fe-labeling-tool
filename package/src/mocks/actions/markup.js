const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
