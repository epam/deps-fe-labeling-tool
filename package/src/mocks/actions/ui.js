const setMarkupObjectsFilter = jest.fn((filter) => ({
  type: 'MOCK_MARKUP_OBJECTS_FILTER_ACTION',
  payload: filter
}))

const setAssignToFieldsFilter = jest.fn((filter) => ({
  type: 'MOCK_ASSIGN_TO_FIELDS_FILTER_ACTION',
  payload: filter
}))

const setExpandedListKeys = jest.fn((key) => ({
  type: 'MOCK_SET_EXPANDED_LIST_KEYS_ACTION',
  payload: key
}))

const toggleExpandedListKey = jest.fn((key) => ({
  type: 'MOCK_TOGGLE_EXPANDED_LIST_KEY_ACTION',
  payload: key
}))

const addTemporaryFieldIndex = jest.fn((key) => ({
  type: 'MOCK_ADD_TEMPORARY_FIELD_INDEX',
  payload: key
}))

const deleteTemporaryFieldIndex = jest.fn((key) => ({
  type: 'MOCK_DELETE_TEMPORARY_FIELD_INDEX',
  payload: key
}))

const resetDefault = jest.fn(() => ({
  type: 'MOCK_RESET_DEFAULT'
}))

const setActiveSidebar = jest.fn(() => ({
  type: 'MOCK_SET_ACTIVE_SIDEBAR'
}))

const mockUiActions = {
  setExpandedListKeys,
  toggleExpandedListKey,
  setMarkupObjectsFilter,
  setAssignToFieldsFilter,
  addTemporaryFieldIndex,
  deleteTemporaryFieldIndex,
  resetDefault,
  setActiveSidebar
}

export {
  mockUiActions
}
