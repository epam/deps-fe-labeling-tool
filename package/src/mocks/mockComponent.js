const mockComponent = (componentName, namedExport = true, fn = () => { }) => {
  Object.defineProperty(fn, 'name', { value: componentName })
  return !namedExport ? fn : { [componentName]: fn }
}

export {
  mockComponent
}
