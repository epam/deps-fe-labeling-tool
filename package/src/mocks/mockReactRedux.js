const mockDispatch = jest.fn((action) => action)

const createMockMapStateToProps = (mapStateToProps) => {
  if (!mapStateToProps) {
    return mapStateToProps
  }

  return (state, ownProps) => ({
    props: mapStateToProps(state, ownProps)
  })
}

const getDispatchFunctionFormShorthandObject = (actionsObj) => (dispatch) => (
  Object.keys(actionsObj).reduce(
    (props, actionKey) => ({
      ...props,
      [actionKey]: jest.fn(
        (...args) => dispatch(actionsObj[actionKey].apply(undefined, args))
      )
    }),
    {}
  )
)

const createMockMapDispatchToProps = (mapDispatchToProps) => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps
  }

  const mappingFunction = typeof mapDispatchToProps === 'function'
    ? mapDispatchToProps
    : getDispatchFunctionFormShorthandObject(mapDispatchToProps)

  return (dispatch, ownProps) => {
    dispatch = dispatch || mockDispatch

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    }
  }
}

const createMockMergeProps = (mergeProps) => {
  if (!mergeProps) {
    return mergeProps
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: mockDispatch }

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    }
  }
}

const mockReactRedux = {
  connect: (mapStateToProps, mapDispatchToProps, mergeProps) => (WrappedComponent) => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    WrappedComponent
  }),
  batch: (fn) => Promise.resolve(fn),
  useDispatch: jest.fn(() => mockDispatch),
  useSelector: jest.fn((selector) => selector())
}

export {
  mockReactRedux
}
