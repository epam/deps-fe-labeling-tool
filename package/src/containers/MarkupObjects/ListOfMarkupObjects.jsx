import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { toggleExpandedListKey } from '@/actions/ui'
import { FieldType } from '@/enums/FieldType'
import { fieldShape } from '@/models/Field'
import { labelShape, Label } from '@/models/Label'
import { tableShape } from '@/models/Table'
import { expandedListKeysSelector } from '@/selectors/ui'
import { replaceAll } from '@/utils/string'
import { ListContentWrapper, ListName, Collapse } from './ListOfMarkupObjects.styles'

const RESOURCE_LIST_NAME = 'List of {0}s'

const getUniqueSets = (markupObjects) => {
  const uniqueSets = []

  markupObjects.forEach((currentLabel) => {
    const hasDuplicate = uniqueSets.some((set) => set.includes(currentLabel))
    if (hasDuplicate) {
      return
    }

    const sameLabels = Label.getSameLabels(currentLabel, markupObjects)
    uniqueSets.push(sameLabels)
  })

  return uniqueSets
}

const ListOfMarkupObjects = ({
  markupObjects,
  field,
  renderMapper,
  expandedListKeys,
  toggleExpandedListKey
}) => {
  const renderer = renderMapper[field.fieldMeta.baseType]

  const onChangeHandler = useCallback(() => {
    toggleExpandedListKey(field.code)
  }, [field.code, toggleExpandedListKey])

  const setsOfLabels = getUniqueSets(markupObjects)

  return (
    <Collapse
      activeKey={expandedListKeys.find((key) => key === field.code)}
      onChange={onChangeHandler}
      header={
        <ListName>
          {replaceAll(RESOURCE_LIST_NAME, field.name)}
        </ListName>
      }
      collapseId={field.code}
    >
      <ListContentWrapper>
        {
          field.fieldMeta.baseType === FieldType.PAIR
            ? renderer(markupObjects, field)
            : setsOfLabels.map((labels, i) => renderer(labels, field, i))
        }
      </ListContentWrapper>
    </Collapse>
  )
}

ListOfMarkupObjects.propTypes = {
  markupObjects: PropTypes.arrayOf(
    PropTypes.oneOfType([
      labelShape,
      tableShape
    ])
  ),
  field: fieldShape,
  toggleExpandedListKey: PropTypes.func.isRequired,
  renderMapper: PropTypes.shape({
    [PropTypes.string]: PropTypes.func
  }).isRequired,
  expandedListKeys: PropTypes.arrayOf(PropTypes.string)
}

const mapStateToProps = (state) => ({
  expandedListKeys: expandedListKeysSelector(state)
})

const mapDispatchToProps = {
  toggleExpandedListKey
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(ListOfMarkupObjects)

export {
  ConnectedComponent as ListOfMarkupObjects
}
