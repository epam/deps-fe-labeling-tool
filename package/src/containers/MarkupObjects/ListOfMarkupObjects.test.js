import { mockUiActions } from '@/mocks/actions/ui'
import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockMarkupSelectors } from '@/mocks/selectors/markup'
import { mockUiSelectors } from '@/mocks/selectors/ui'
import React from 'react'
import { shallow } from 'enzyme'
import { toggleExpandedListKey } from '@/actions/ui'
import { FieldType } from '@/enums/FieldType'
import { Field } from '@/models/Field'
import { ListFieldMeta } from '@/models/FieldMeta'
import { ListOfMarkupObjects } from './ListOfMarkupObjects'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/actions/ui', () => mockUiActions)
jest.mock('@/selectors/markup', () => mockMarkupSelectors)
jest.mock('./PrimitiveMarkupObject', () => mockComponent('PrimitiveMarkupObject'))
jest.mock('@/selectors/ui', () => mockUiSelectors)

const mockField = new Field(
  'field_code',
  'Field Name',
  FieldType.LIST,
  new ListFieldMeta(FieldType.STRING)
)

const { WrappedComponent, mapStateToProps, mapDispatchToProps } = ListOfMarkupObjects

describe('Container: ListOfMarkupObjects', () => {
  describe('mapDispatchToProps', () => {
    const { props } = mapDispatchToProps()

    it('should call for toggleExpandedListKey action from props', () => {
      props.toggleExpandedListKey()
      expect(toggleExpandedListKey).toHaveBeenCalledTimes(1)
    })
  })

  describe('mapStateToProps', () => {
    it('should pass state as props to WrappedComponent', () => {
      const { props } = mapStateToProps()
      const expectedProps = {
        expandedListKeys: mockUiSelectors.expandedListKeysSelector()
      }
      expect(props).toEqual(expectedProps)
    })
  })

  describe('ConnectedComponent', () => {
    let defaultProps
    let wrapper

    beforeEach(() => {
      defaultProps = {
        field: mockField,
        markupObjects: mockMarkupSelectors.pageLabelsSelector(),
        renderMapper: {
          [FieldType.STRING]: jest.fn()
        },
        expandedListKeys: mockUiSelectors.expandedListKeysSelector(),
        ...mapDispatchToProps().props
      }

      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render correct layout', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call toggleExpandedListKey with correct arg when calling onChange', () => {
      wrapper.props().onChange()

      expect(toggleExpandedListKey).toHaveBeenCalled()
    })
  })
})
