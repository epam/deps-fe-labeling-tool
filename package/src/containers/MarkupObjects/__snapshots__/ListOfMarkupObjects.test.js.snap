// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: ListOfMarkupObjects ConnectedComponent should render correct layout 1`] = `
<ListOfMarkupObjectsstyles__MarkupCollapse
  collapseId="field_code"
  header={
    <ListOfMarkupObjectsstyles__ListName>
      List of Field Names
    </ListOfMarkupObjectsstyles__ListName>
  }
  onChange={[Function]}
>
  <ListOfMarkupObjectsstyles__ListContentWrapper />
</ListOfMarkupObjectsstyles__MarkupCollapse>
`;
